### PR TITLE
Major fixes for C8DASR recipe: GSS and Channel clustering

### PR DESCRIPTION
**Pull request by popcornell**
This PR fixes: 
1. GSS, now performed freq-chunk wise, better performance and less memory
2. Channel Clustering failing when zeros are in the input file.

### CODE REVIEW

File: scripts/chime7/pipeline/gss_process/run_gss_process.py [hunk 1]
```diff
@@ -34,7 +34,7 @@
 def convert_diar_results_to_falign(scenarios: list, diarization_dir: str, output_dir: str, subsets: list = ['dev']):
     # Assumption:
     # Output of diarization is organized in 3 subdirectories, with each subdirectory corresponding to one scenario (chime6, dipco, mixer6)
-    diar_json_dir = os.path.join(diarization_dir, "pred_jsons_T")
+    diar_json_dir = os.path.join(diarization_dir, "pred_jsons_T0.55")
```
Comment (popcornell) [inline scripts/chime7/pipeline/gss_process/run_gss_process.py]: @tango4j why is the name different on my side ?

File: nemo/collections/asr/modules/audio_modules.py [hunk 2]
```diff
@@ -969,7 +969,7 @@ def forward(self, input: torch.Tensor, activity: torch.Tensor) -> torch.Tensor:
 
         return gamma
 
-
+#FIXME DUPLICATED
 class MaskEstimatorGSS(torch.nn.Module):
```
Comment (popcornell) [inline nemo/collections/asr/modules/audio_modules.py:973]: MaskEstimationGSS is duplicated on this file

**Comment by popcornell** [comment]
I fixed GSS to be freq-wise. 
On DiPCo dev set session 26 it improves performance from 0.43 to 0.38. So quite a bit.

File: nemo/collections/asr/parts/preprocessing/segment.py [hunk 3]
```diff
@@ -358,9 +358,15 @@ def from_file_list(
...
                 # Check the dimensions match
                 if len(a_samples) != len(samples):
-                    raise RuntimeError(
-                        f'Loaded samples need to have identical length: {a_samples.shape} != {samples.shape}'
-                    )
+
+                    logging.warning(f"Loaded samples for {a_file}, offset {offset}, duration {duration} have different length from previously loaded ones: {a_samples.shape} != {samples.shape}")
+                    logging.warning("Using zero padding !")
+                    maxlen = max(len(a_samples), len(samples))
+                    a_samples = np.pad(a_samples, ((0, maxlen - len(a_samples)), (0, 0)))
```
Comment (popcornell) [inline nemo/collections/asr/parts/preprocessing/segment.py:365]: I have to do some padding here because of the fact that some files in CHiME-6 are longer than others.

File: nemo/collections/asr/parts/utils/channel_clustering.py [hunk 4]
```diff
@@ -111,15 +111,15 @@ def channel_cluster(
     oracle_num_channels: int = -1,
     max_rp_threshold: float = 0.5,
     max_num_channels: int = 7,
-    min_num_channels: int = 1,
+    min_num_channels: int = 3,
```
Comment (popcornell) [inline nemo/collections/asr/parts/utils/channel_clustering.py:114]: This prevents the channel clustering collapsing when the input is all zeros

File: scripts/chime7/pipeline/chime_config.yaml [hunk 5]
```diff
@@ -16,7 +16,7 @@ subsets:
 
 num_workers: 16
 sample_rate: 16000
-batch_size: 8  # batch size for diarization and ASR
+batch_size: 1  # batch size for diarization and ASR
```
Comment (popcornell) [inline scripts/chime7/pipeline/chime_config.yaml]: @tango4j we need separate batch size for major steps in the pipeline. 
ASR needs batch size 1 or I get OOM on dipco.

File: scripts/chime7/pipeline/gss_process/run_gss_process.py (hunk 1 above)
Comment (popcornell) [reply scripts/chime7/pipeline/gss_process/run_gss_process.py]: The name sometimes is T0.55 (dipco) sometimes is T (CHiME-6)

Comment (tango4j) [reply scripts/chime7/pipeline/gss_process/run_gss_process.py]: This should be "pred_jsons_T". everything else is mistake.

File: scripts/chime7/pipeline/chime_config.yaml (hunk 5 above)
Comment (tango4j) [reply scripts/chime7/pipeline/chime_config.yaml]: This should be kept as batch_size:8. 
The length of the to-be-GSS segments should be controlled by min_seg_length.
I guess this mechanism is not working for some reason

File: nemo/collections/asr/modules/audio_modules.py [hunk 6]
```diff
@@ -1668,16 +1668,18 @@ def estimate_filter(self, Q: torch.Tensor, R: torch.Tensor) -> torch.Tensor:
...
             # Use Cholesky decomposition
             QL = torch.linalg.cholesky(Q)
             # Forward and backward substitution
             G = torch.linalg.solve_triangular(QL, R, upper=False)
             G = torch.linalg.solve_triangular(QL.conj().transpose(-2, -1), G, upper=True)
         except torch.linalg.LinAlgError as e:
-            # Fallback to linsolve
-            logging.warning('Cholesky failed with exception: %s', e)
-            logging.info('Fallback to linsolve')
+            fail = True
```
Comment (popcornell) [inline nemo/collections/asr/modules/audio_modules.py:1679]: @tango4j this caused memory leakage in GSS processing. 
see https://pytorch.org/docs/stable/notes/faq.html#my-out-of-memory-exception-handler-can-t-allocate-memory

File: scripts/chime7/pipeline/run_full_pipeline.py [hunk 7]
```diff
@@ -22,26 +22,61 @@
 
 from nemo.core.config import hydra_runner
 from nemo.utils import logging
+from functools import partial
+from functools import partial
+
+def run_stage(stage_num, start_stage=-1, stop_stage=np.inf, skip_stages=None):
```
Comment (popcornell) [inline scripts/chime7/pipeline/run_full_pipeline.py:28]: @tango4j simple code for skipping stage

File: scripts/chime7/pipeline/chime_config.yaml [hunk 8]
```diff
@@ -182,7 +199,7 @@ asr:
 
   # General configs
   output_filename: null
-  batch_size: ${batch_size}  # use a smaller batch if beam size is large
+  batch_size: ${asr_batch_size}  # use a smaller batch if beam size is large
```
Comment (popcornell) [inline scripts/chime7/pipeline/chime_config.yaml:202]: added ASR batch size

File: scripts/chime7/pipeline/gss_process/run_gss_process.py (hunk 1 above)
Comment (popcornell) [reply scripts/chime7/pipeline/gss_process/run_gss_process.py]: Nevermind this problem seems has been solved

File: scripts/chime7/pipeline/chime_config.yaml (hunk 5 above)
Comment (popcornell) [reply scripts/chime7/pipeline/chime_config.yaml]: Added

File: scripts/chime7/pipeline/run_full_pipeline.py (hunk 7 above)
Comment (tango4j) [reply scripts/chime7/pipeline/run_full_pipeline.py:28]: np is not imported. I am adding it.

File: nemo/collections/asr/parts/preprocessing/segment.py (hunk 3 above)
Comment (popcornell) [reply nemo/collections/asr/parts/preprocessing/segment.py:365]: @tango4j padding is here

**Review by tango4j — APPROVED** [review]
Approving
